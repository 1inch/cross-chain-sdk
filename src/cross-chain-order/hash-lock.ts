import {id} from 'ethers'
import {getBytesCount} from '@1inch/byte-utils'
import assert from 'assert'

export class HashLock {
    public static Web3Type = 'bytes32'

    private readonly value: string

    protected constructor(val: string) {
        this.value = val
    }

    /**
     * Create HashLock from keccak256 hash of secret
     */
    public static fromSecret(secret: string): HashLock {
        return new HashLock(id(secret))
    }

    public static fromString(hashLock: string): HashLock {
        assert(
            getBytesCount(hashLock) === 32n,
            'HashLock string must be result of keccak256'
        )

        return new HashLock(hashLock)
    }

    public toString(): string {
        return this.value
    }
}
