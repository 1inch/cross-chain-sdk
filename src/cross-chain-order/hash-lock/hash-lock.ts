import {keccak256, solidityPackedKeccak256} from 'ethers'
import {SimpleMerkleTree} from '@openzeppelin/merkle-tree'
import {BitMask, BN, getBytesCount, isHexBytes} from '@1inch/byte-utils'
import assert from 'assert'

export class HashLock {
    public static Web3Type = 'bytes32'

    private readonly value: string

    protected constructor(val: string) {
        this.value = val
    }

    public static hashSecret(secret: string): string {
        assert(
            isHexBytes(secret) && getBytesCount(secret) === 32n,
            'secret length must be 32 bytes hex encoded'
        )

        return keccak256(secret)
    }

    public static getMerkleLeaves(secrets: string[]): MerkleLeaf[] {
        return HashLock.getMerkleLeavesFromSecretHashes(
            secrets.map(HashLock.hashSecret)
        )
    }

    public static getMerkleLeavesFromSecretHashes(
        secretHashes: string[]
    ): MerkleLeaf[] {
        return secretHashes.map(
            (s, idx) =>
                solidityPackedKeccak256(
                    ['uint64', 'bytes32'],
                    [idx, s]
                ) as MerkleLeaf
        )
    }

    public static getProof(leaves: string[], idx: number): MerkleLeaf[] {
        return SimpleMerkleTree.of(leaves).getProof(idx) as MerkleLeaf[]
    }

    public static fromString(value: string): HashLock {
        assert(
            isHexBytes(value) && getBytesCount(value) === 32n,
            'HashLock value must be bytes32 hex encoded'
        )

        return new HashLock(value)
    }

    /**
     * Create HashLock from keccak256 hash of secret
     */
    public static forSingleFill(secret: string): HashLock {
        return new HashLock(HashLock.hashSecret(secret))
    }

    public static forMultipleFills(leaves: MerkleLeaf[]): HashLock {
        assert(
            leaves.length > 2,
            'leaves array must be greater than 2. Or use HashLock.forSingleFill'
        )
        const root = SimpleMerkleTree.of(leaves).root
        const rootWithCount = BN.fromHex(root).setMask(
            new BitMask(240n, 256n),
            BigInt(leaves.length - 1)
        )

        return new HashLock(rootWithCount.toHex(64))
    }

    /**
     * Only use if HashLockInfo is for multiple fill order
     * Otherwise garbage will be returned
     *
     */
    public getPartsCount(): bigint {
        return new BN(BigInt(this.value)).getMask(new BitMask(240n, 256n)).value
    }

    public toString(): string {
        return this.value
    }

    public eq(other: HashLock): boolean {
        return this.value === other.value
    }
}

export type MerkleLeaf = string & {_tag: 'MerkleLeaf'}
