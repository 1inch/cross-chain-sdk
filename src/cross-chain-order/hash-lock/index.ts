export {HashLock} from './hash-lock'
