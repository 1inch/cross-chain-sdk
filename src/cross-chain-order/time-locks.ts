import {BitMask, BN, UINT_32_MAX} from '@1inch/byte-utils'
import assert from 'assert'

// todo check intervals and add docs
export class TimeLocks {
    static Web3Type = 'uint256'

    protected constructor(
        public readonly deployedAt: bigint,
        private readonly _srcWithdrawal: bigint,
        private readonly _srcPublicWithdrawal: bigint,
        private readonly _srcCancellation: bigint,
        private readonly _srcPublicCancellation: bigint,
        private readonly _dstWithdrawal: bigint,
        private readonly _dstPublicWithdrawal: bigint,
        private readonly _dstCancellation: bigint
    ) {
        assert(
            deployedAt <= UINT_32_MAX,
            'deployedAt can not be > uint32 max value'
        )

        assert(
            _srcWithdrawal <= UINT_32_MAX,
            'srcWithdrawal can not be > uint32 max value'
        )
        assert(
            _srcPublicWithdrawal <= UINT_32_MAX,
            'srcPublicWithdrawal can not be > uint32 max value'
        )
        assert(
            _srcCancellation <= UINT_32_MAX,
            'srcCancellation can not be > uint32 max value'
        )
        assert(
            _srcPublicCancellation <= UINT_32_MAX,
            'srcPublicCancellation can not be > uint32 max value'
        )
        assert(
            _dstWithdrawal <= UINT_32_MAX,
            'dstWithdrawal can not be > uint32 max value'
        )
        assert(
            _dstPublicWithdrawal <= UINT_32_MAX,
            'dstPublicWithdrawal can not be > uint32 max value'
        )
        assert(
            _dstCancellation <= UINT_32_MAX,
            'dstCancellation can not be > uint32 max value'
        )
    }

    public get srcWithdrawal(): bigint {
        return this.deployedAt + this._srcWithdrawal
    }

    public get srcPublicWithdrawal(): bigint {
        return this.deployedAt + this._srcPublicWithdrawal
    }

    public get srcCancellation(): bigint {
        return this.deployedAt + this._srcCancellation
    }

    public get srcPublicCancellation(): bigint {
        return this.deployedAt + this._srcPublicCancellation
    }

    public get dstWithdrawal(): bigint {
        return this.deployedAt + this._dstWithdrawal
    }

    public get dstPublicWithdrawal(): bigint {
        return this.deployedAt + this._dstPublicWithdrawal
    }

    public get dstCancellation(): bigint {
        return this.deployedAt + this._dstCancellation
    }

    public static new(params: {
        deployedAt: bigint
        srcWithdrawal: bigint
        srcPublicWithdrawal: bigint
        srcCancellation: bigint
        srcPublicCancellation: bigint
        dstWithdrawal: bigint
        dstPublicWithdrawal: bigint
        dstCancellation: bigint
    }): TimeLocks {
        return new TimeLocks(
            params.deployedAt,
            params.srcWithdrawal,
            params.srcPublicWithdrawal,
            params.srcCancellation,
            params.srcPublicCancellation,
            params.dstWithdrawal,
            params.dstPublicWithdrawal,
            params.dstCancellation
        )
    }

    public static fromBigInt(val: bigint): TimeLocks {
        const valBN = new BN(val)

        const params = Array.from({length: 8}).map((_, i) => {
            return valBN.getMask(
                new BitMask(BigInt(i) * 32n, BigInt(i + 1) * 32n)
            ).value
        }) as [bigint, bigint, bigint, bigint, bigint, bigint, bigint, bigint] // ts cannot infer that we have exactly 8 elements in the array after map

        return new TimeLocks(...params)
    }

    public build(): bigint {
        return [
            this.deployedAt,
            this._srcWithdrawal,
            this._srcPublicWithdrawal,
            this._srcCancellation,
            this._srcPublicCancellation,
            this._dstWithdrawal,
            this._dstPublicWithdrawal,
            this._dstCancellation
        ].reduceRight((acc, el) => (acc << 32n) | el)
    }
}
