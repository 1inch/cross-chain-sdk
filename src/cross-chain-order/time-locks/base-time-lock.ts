import {UINT_32_MAX} from '@1inch/byte-utils'
import assert from 'assert'

export abstract class BaseTimeLock {
    static DEFAULT_RESCUE_DELAY = 604800n // 7 days

    protected constructor(public readonly deployedAt: bigint) {
        assert(deployedAt !== 0n, 'deployedAt must be > 0n')

        assert(
            deployedAt <= UINT_32_MAX,
            'deployedAt can not be > uint32 max value'
        )
    }

    public getRescueStart(
        rescueDelay = BaseTimeLock.DEFAULT_RESCUE_DELAY
    ): bigint {
        return this.deployedAt + rescueDelay
    }
}
