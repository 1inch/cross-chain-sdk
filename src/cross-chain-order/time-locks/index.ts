export {TimeLocks} from './time-locks'
export {SrcTimeLocks} from './src-time-locks'
export {DstTimeLocks} from './dst-time-locks'
