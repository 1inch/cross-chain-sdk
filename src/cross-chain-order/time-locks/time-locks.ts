import {BitMask, BN, UINT_32_MAX} from '@1inch/byte-utils'
import assert from 'assert'
import {SrcTimeLocks} from './src-time-locks'
import {DstTimeLocks} from './dst-time-locks'

/**
 * Contains the duration of each stage of swap for source and destination chain
 *
 * Source chain intervals layout
 * | finality lock | private withdrawal | public withdrawal | private cancellation | public cancellation |
 * ^deployedAt
 *
 * Destination chain intervals layout
 * | finality lock | private withdrawal | public withdrawal | private cancellation |
 * ^deployedAt
 *
 * @see SrcTimeLocks
 * @see DstTimeLocks
 */
export class TimeLocks {
    static DEFAULT_RESCUE_DELAY = 604800n // 7 days

    static Web3Type = 'uint256'

    protected constructor(
        public readonly deployedAt: bigint,
        private readonly _srcWithdrawal: bigint,
        private readonly _srcPublicWithdrawal: bigint,
        private readonly _srcCancellation: bigint,
        private readonly _srcPublicCancellation: bigint,
        private readonly _dstWithdrawal: bigint,
        private readonly _dstPublicWithdrawal: bigint,
        private readonly _dstCancellation: bigint
    ) {
        assert(
            deployedAt <= UINT_32_MAX,
            'deployedAt can not be > uint32 max value'
        )

        assert(
            _srcWithdrawal <= UINT_32_MAX,
            'srcWithdrawal can not be > uint32 max value'
        )
        assert(
            _srcWithdrawal < _srcPublicWithdrawal,
            'srcWithdrawal can not be >= srcPublicWithdrawal'
        )

        assert(
            _srcPublicWithdrawal <= UINT_32_MAX,
            'srcPublicWithdrawal can not be > uint32 max value'
        )
        assert(
            _srcPublicWithdrawal < _srcCancellation,
            'srcPublicWithdrawal can not be >= srcCancellation'
        )

        assert(
            _srcCancellation <= UINT_32_MAX,
            'srcCancellation can not be > uint32 max value'
        )
        assert(
            _srcCancellation < _srcPublicCancellation,
            'srcCancellation can not be >= srcPublicCancellation'
        )

        assert(
            _srcPublicCancellation <= UINT_32_MAX,
            'srcPublicCancellation can not be > uint32 max value'
        )

        assert(
            _dstWithdrawal <= UINT_32_MAX,
            'dstWithdrawal can not be > uint32 max value'
        )
        assert(
            _dstWithdrawal < _dstPublicWithdrawal,
            'dstWithdrawal can not be >= dstPublicWithdrawal'
        )

        assert(
            _dstPublicWithdrawal <= UINT_32_MAX,
            'dstPublicWithdrawal can not be > uint32 max value'
        )
        assert(
            _dstPublicWithdrawal < _dstCancellation,
            'dstPublicWithdrawal can not be >= dstCancellation'
        )

        assert(
            _dstCancellation <= UINT_32_MAX,
            'dstCancellation can not be > uint32 max value'
        )
    }

    public static new(params: {
        /**
         * Network: Source
         * Delay from `deployedAt` at which ends `finality lock` and starts `private withdrawal` */
        srcWithdrawal: bigint
        /**
         * Network: Source
         * Delay from `deployedAt` at which ends `private withdrawal` and starts `public withdrawal` */
        srcPublicWithdrawal: bigint
        /**
         * Network: Source
         * Delay from `deployedAt` at which ends `public withdrawal` and starts `private cancellation` */
        srcCancellation: bigint
        /**
         * Network: Source
         * Delay from `deployedAt` at which ends `private cancellation` and starts `public cancellation` */
        srcPublicCancellation: bigint
        /**
         * Network: Destination
         * Delay from `deployedAt` at which ends `finality lock` and starts `private withdrawal` */
        dstWithdrawal: bigint
        /**
         * Network: Destination
         * Delay from `deployedAt` at which ends `private withdrawal` and starts `public withdrawal` */
        dstPublicWithdrawal: bigint
        /**
         * Network: Destination
         * Delay from `deployedAt` at which ends `public withdrawal` and starts `private cancellation` */
        dstCancellation: bigint
    }): TimeLocks {
        return new TimeLocks(
            0n,
            params.srcWithdrawal,
            params.srcPublicWithdrawal,
            params.srcCancellation,
            params.srcPublicCancellation,
            params.dstWithdrawal,
            params.dstPublicWithdrawal,
            params.dstCancellation
        )
    }

    public static fromBigInt(val: bigint): TimeLocks {
        const valBN = new BN(val)

        const params = Array.from({length: 8}).map((_, i) => {
            return valBN.getMask(
                new BitMask(BigInt(i) * 32n, BigInt(i + 1) * 32n)
            ).value
        }) as [bigint, bigint, bigint, bigint, bigint, bigint, bigint, bigint] // ts cannot infer that we have exactly 8 elements in the array after map

        return new TimeLocks(...params)
    }

    public build(): bigint {
        return [
            this.deployedAt,
            this._srcWithdrawal,
            this._srcPublicWithdrawal,
            this._srcCancellation,
            this._srcPublicCancellation,
            this._dstWithdrawal,
            this._dstPublicWithdrawal,
            this._dstCancellation
        ].reduceRight((acc, el) => (acc << 32n) | el)
    }

    public toSrcTimeLocks(deployedAt = this.deployedAt): SrcTimeLocks {
        return SrcTimeLocks.new({
            deployedAt,
            withdrawal: this._srcWithdrawal,
            publicWithdrawal: this._srcPublicWithdrawal,
            cancellation: this._srcCancellation,
            publicCancellation: this._srcPublicCancellation
        })
    }

    public toDstTimeLocks(deployedAt = this.deployedAt): DstTimeLocks {
        return DstTimeLocks.new({
            deployedAt,
            withdrawal: this._dstWithdrawal,
            publicWithdrawal: this._dstPublicWithdrawal,
            cancellation: this._dstCancellation
        })
    }
}
