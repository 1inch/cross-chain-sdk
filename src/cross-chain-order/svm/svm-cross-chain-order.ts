import {BitMask, UINT_32_MAX, UINT_64_MAX} from '@1inch/byte-utils'
import {AuctionCalculator, randBigInt} from '@1inch/fusion-sdk'
import {keccak256} from 'ethers'
import {utils} from '@coral-xyz/anchor'
import assert from 'assert'
import {ResolverCancellationConfig} from './resolver-cancellation-config'
import {SolanaDetails, SolanaExtra, SolanaEscrowParams} from './types'
import {SvmSrcEscrowFactory} from '../../contracts'
import {hashForSolana} from '../../domains/auction-details/hasher'
import {uint256BorchSerialized} from '../../utils/numbers/uint256-borsh-serialized'
import {uintAsBeBytes} from '../../utils/numbers/uint-as-be-bytes'
import {AddressLike, EvmAddress, SolanaAddress} from '../../domains/addresses'
import {isSupportedChain, NetworkEnum, SupportedChain} from '../../chains'
import {HashLock} from '../../domains/hash-lock'
import {TimeLocks} from '../../domains/time-locks'
import {BaseOrder} from '../base-order'
import {assertUInteger} from '../../utils'
import {AuctionDetails, AuctionPoint} from '../../domains/auction-details'
import {injectTrackCode} from '../source-track'

export type SolanaOrderJSON = {
    orderInfo: {
        srcToken: string // base58 solana address
        dstToken: string // evm address
        maker: string // base58 solana address
        srcAmount: string // u64 bigint
        minDstAmount: string // u64 bigint
        receiver: string // evm address
    }
    escrowParams: {
        hashLock: string // 32bytes hex
        srcChainId: NetworkEnum.SOLANA
        dstChainId: number
        srcSafetyDeposit: string // u64 bigint
        dstSafetyDeposit: string // u64 bigint
        timeLocks: string // u256 bigint
    }
    details: {
        auction: {
            startTime: string
            duration: string
            initialRateBump: number
            points: AuctionPoint[]
        }
    }
    extra: {
        srcAssetIsNative: boolean
        orderExpirationDelay: string // bigint
        resolverCancellationConfig: {
            maxCancellationPremium: string // u64 bigint
            cancellationAuctionDuration: number
        }
        source: string
        allowMultipleFills: boolean
        salt: string // u32 bigint
    }
}

export type OrderInfoData = {
    srcToken: SolanaAddress
    dstToken: EvmAddress
    maker: SolanaAddress
    srcAmount: bigint // u64
    minDstAmount: bigint // u64
    receiver: EvmAddress
}

export class SvmCrossChainOrder extends BaseOrder<
    SolanaAddress,
    SolanaOrderJSON
> {
    private static TRACK_CODE_MASK = new BitMask(32n, 64n)

    private static DefaultExtra = {
        orderExpirationDelay: 12n,
        allowMultipleFills: true,
        source: 'sdk',
        resolverCancellationConfig: ResolverCancellationConfig.ALMOST_ZERO // to enable cancellation by resolver
    }

    private readonly orderConfig: {
        srcToken: SolanaAddress
        dstToken: AddressLike
        maker: SolanaAddress
        receiver: EvmAddress
        srcAmount: bigint // u64
        minDstAmount: bigint // u64
        deadline: number // u32
        salt: bigint // u64

        // ---extra---
        srcAssetIsNative: boolean
        // dst asset is native in case dstToken.isZero
        resolverCancellationConfig: ResolverCancellationConfig
        allowMultipleFills: boolean
        orderExpirationDelay: bigint
        source: string
    }

    private readonly details: SolanaDetails

    private readonly escrowParams: SolanaEscrowParams

    private readonly encoder = new TextEncoder()

    private constructor(
        orderInfo: OrderInfoData,
        escrowParams: SolanaEscrowParams,
        details: SolanaDetails,
        extra: SolanaExtra
    ) {
        assert(
            isSupportedChain(escrowParams.srcChainId),
            `Not supported chain ${escrowParams.srcChainId}`
        )
        assert(
            isSupportedChain(escrowParams.dstChainId),
            `Not supported chain ${escrowParams.dstChainId}`
        )
        assert(
            escrowParams.srcChainId !== escrowParams.dstChainId,
            'Chains must be different'
        )

        super()

        const orderExpirationDelay =
            extra.orderExpirationDelay ??
            SvmCrossChainOrder.DefaultExtra.orderExpirationDelay
        assertUInteger(orderExpirationDelay)

        const deadline =
            details.auction.startTime +
            details.auction.duration +
            orderExpirationDelay

        assertUInteger(deadline, UINT_32_MAX)
        assertUInteger(orderInfo.srcAmount, UINT_64_MAX)
        assertUInteger(orderInfo.minDstAmount, UINT_64_MAX)

        // todo more asserts

        const source = extra.source ?? SvmCrossChainOrder.DefaultExtra.source
        const salt = injectTrackCode(
            extra.salt ?? randBigInt(UINT_32_MAX),
            source,
            SvmCrossChainOrder.TRACK_CODE_MASK
        )

        const resolverCancellationConfig =
            extra.resolverCancellationConfig ||
            SvmCrossChainOrder.DefaultExtra.resolverCancellationConfig

        this.details = details
        this.escrowParams = escrowParams
        this.orderConfig = {
            ...orderInfo,
            source,
            salt,
            allowMultipleFills:
                extra.allowMultipleFills ??
                SvmCrossChainOrder.DefaultExtra.allowMultipleFills,
            srcAssetIsNative: extra.srcAssetIsNative || false,
            deadline: Number(deadline),
            resolverCancellationConfig: resolverCancellationConfig,
            orderExpirationDelay
        }
    }

    public get auction(): AuctionDetails {
        return this.details.auction
    }

    public get salt(): bigint {
        return this.orderConfig.salt
    }

    public get resolverCancellationConfig(): ResolverCancellationConfig {
        return this.orderConfig.resolverCancellationConfig
    }

    public get hashLock(): HashLock {
        return this.escrowParams.hashLock
    }

    public get timeLocks(): TimeLocks {
        return this.escrowParams.timeLocks
    }

    public get srcSafetyDeposit(): bigint {
        return this.escrowParams.srcSafetyDeposit
    }

    public get dstSafetyDeposit(): bigint {
        return this.escrowParams.dstSafetyDeposit
    }

    public get dstChainId(): SupportedChain {
        return this.escrowParams.dstChainId
    }

    public get maker(): SolanaAddress {
        return this.orderConfig.maker
    }

    public get makerAsset(): SolanaAddress {
        return this.orderConfig.srcToken
    }

    public get takerAsset(): AddressLike {
        return this.orderConfig.dstToken
    }

    public get makingAmount(): bigint {
        return this.orderConfig.srcAmount
    }

    public get takingAmount(): bigint {
        return this.orderConfig.minDstAmount
    }

    public get receiver(): EvmAddress {
        return this.orderConfig.receiver
    }

    public get deadline(): bigint {
        return BigInt(this.orderConfig.deadline)
    }

    public get auctionStartTime(): bigint {
        return this.details.auction.startTime
    }

    public get auctionEndTime(): bigint {
        return this.auctionStartTime + this.details.auction.duration
    }

    public get partialFillAllowed(): boolean {
        return this.orderConfig.allowMultipleFills
    }

    public get multipleFillsAllowed(): boolean {
        return this.orderConfig.allowMultipleFills
    }

    public get srcAssetIsNative(): boolean {
        return this.orderConfig.srcAssetIsNative
    }

    static new(
        orderInfo: OrderInfoData,
        escrowParams: SolanaEscrowParams,
        details: SolanaDetails,
        extra: Omit<SolanaExtra, 'srcAssetIsNative'>
    ): SvmCrossChainOrder {
        return new SvmCrossChainOrder(
            {
                ...orderInfo,
                srcToken: orderInfo.srcToken.isNative()
                    ? SolanaAddress.WRAPPED_NATIVE
                    : orderInfo.srcToken
            },
            escrowParams,
            details,
            {
                ...extra,
                srcAssetIsNative: orderInfo.srcToken.isNative()
            }
        )
    }

    static fromJSON(data: SolanaOrderJSON): SvmCrossChainOrder {
        return new SvmCrossChainOrder(
            {
                srcToken: SolanaAddress.fromString(data.orderInfo.srcToken),
                dstToken: EvmAddress.fromString(data.orderInfo.dstToken),
                srcAmount: BigInt(data.orderInfo.srcAmount),
                minDstAmount: BigInt(data.orderInfo.minDstAmount),
                maker: SolanaAddress.fromString(data.orderInfo.maker),
                receiver: EvmAddress.fromString(data.orderInfo.receiver)
            },
            {
                dstChainId: data.escrowParams.dstChainId,
                dstSafetyDeposit: BigInt(data.escrowParams.dstSafetyDeposit),
                hashLock: HashLock.fromString(data.escrowParams.hashLock),
                srcChainId: data.escrowParams.srcChainId,
                srcSafetyDeposit: BigInt(data.escrowParams.srcSafetyDeposit),
                timeLocks: TimeLocks.fromBigInt(
                    BigInt(data.escrowParams.timeLocks)
                )
            },
            {
                auction: AuctionDetails.fromJSON({
                    ...data.details.auction,
                    gasCost: {gasBumpEstimate: '0', gasPriceEstimate: '0'}
                })
            },
            {
                allowMultipleFills: data.extra.allowMultipleFills,
                orderExpirationDelay: BigInt(data.extra.orderExpirationDelay),
                salt: BigInt(data.extra.salt),
                resolverCancellationConfig: new ResolverCancellationConfig(
                    BigInt(
                        data.extra.resolverCancellationConfig
                            .maxCancellationPremium
                    ),
                    data.extra.resolverCancellationConfig.cancellationAuctionDuration
                ),
                source: data.extra.source,
                srcAssetIsNative: data.extra.srcAssetIsNative
            }
        )
    }

    public toJSON(): SolanaOrderJSON {
        const auction = this.auction.toJSON()

        return {
            details: {
                // skip gasCost field
                auction: {
                    duration: auction.duration,
                    initialRateBump: auction.initialRateBump,
                    points: auction.points,
                    startTime: auction.startTime
                }
            },
            orderInfo: {
                srcToken: this.orderConfig.srcToken.toString(),
                dstToken: this.orderConfig.dstToken.toString(),
                maker: this.orderConfig.maker.toString(),
                srcAmount: this.orderConfig.srcAmount.toString(),
                minDstAmount: this.orderConfig.minDstAmount.toString(),
                receiver: this.orderConfig.receiver.toString()
            },
            escrowParams: {
                hashLock: this.hashLock.toString(),
                srcChainId: NetworkEnum.SOLANA,
                dstChainId: this.dstChainId,
                srcSafetyDeposit: this.escrowParams.srcSafetyDeposit.toString(),
                dstSafetyDeposit: this.escrowParams.dstSafetyDeposit.toString(),
                timeLocks: this.timeLocks.build().toString()
            },
            extra: {
                srcAssetIsNative: this.srcAssetIsNative,
                orderExpirationDelay:
                    this.orderConfig.orderExpirationDelay.toString(),
                resolverCancellationConfig:
                    this.resolverCancellationConfig.toJSON(),
                source: this.orderConfig.source,
                allowMultipleFills: this.multipleFillsAllowed,
                // use only last bits because high ones set from source
                salt: (this.salt & UINT_32_MAX).toString()
            }
        }
    }

    public getOrderAccount(
        /**
         * Src escrow factory program id
         */
        programId: SolanaAddress
    ): SolanaAddress {
        return new SvmSrcEscrowFactory(programId).getOrderAccount(
            this.getOrderHashBuffer()
        )
    }

    /**
     * Returns escrow address - owner of ATA where funds stored after fill
     *
     * @see getSrcEscrowATA to get ATA where funds stored
     */
    public getSrcEscrowAddress(
        /**
         * Src escrow factory program id
         */
        programId: SolanaAddress,
        /**
         * Address who fill order and create corresponding escrow
         */
        taker: SolanaAddress,
        /**
         * Hash of corresponding to fill amount secret
         * Can be omitted  for orders where `multipleFillsAllowed` is false
         */
        secretHash = this.hashLock.toBuffer()
    ): SolanaAddress {
        return new SvmSrcEscrowFactory(programId).getEscrowAddress({
            orderHash: this.getOrderHashBuffer(),
            secretHash,
            maker: this.maker,
            taker,
            makerAsset: this.makerAsset,
            makingAmount: this.makingAmount,
            srcSafetyDeposit: this.srcSafetyDeposit
        })
    }

    // /**
    //  * Actual address where funds stored
    //  */
    // public getEscrowATA(
    //     /**
    //      * Src escrow factory
    //      */
    //     srcEscrowProgramId: SolanaAddress,
    //     /**
    //      * TokenProgram or TokenProgram 2022
    //      */
    //     srcMintProgramId: SolanaAddress
    // ): SolanaAddress {
    //     const escrowAddress = this.getEscrowAddress(srcEscrowProgramId)
    //
    //     return getAta(escrowAddress, this.makerAsset, srcMintProgramId)
    // }

    /**
     * @returns order has in base58 encoding
     */
    public getOrderHash(_srcChainId: number): string {
        return utils.bytes.bs58.encode(this.getOrderHashBuffer())
    }

    public getOrderHashBuffer(): Buffer {
        return Buffer.from(
            keccak256(
                Buffer.concat([
                    this.hashLock.toBuffer(),
                    this.maker.toBuffer(),
                    this.makerAsset.toBuffer(),
                    uintAsBeBytes(this.makingAmount, 64),
                    uintAsBeBytes(this.srcSafetyDeposit, 64),
                    uint256BorchSerialized(this.timeLocks.build()),
                    uintAsBeBytes(this.deadline, 32),
                    Buffer.from([Number(this.srcAssetIsNative)]),
                    uint256BorchSerialized(this.takingAmount),
                    hashForSolana(this.auction),
                    uintAsBeBytes(
                        this.resolverCancellationConfig.maxCancellationPremium,
                        64
                    ),
                    uintAsBeBytes(
                        BigInt(
                            this.resolverCancellationConfig
                                .cancellationAuctionDuration
                        ),
                        32
                    ),
                    Buffer.from([Number(this.multipleFillsAllowed)]),
                    uintAsBeBytes(this.salt, 64)
                ])
            ).slice(2),
            'hex'
        )
    }

    public getCalculator(): AuctionCalculator {
        const details = this.details.auction

        return new AuctionCalculator(
            details.startTime,
            details.duration,
            details.initialRateBump,
            details.points,
            0n // no taker fee
        )
    }
}
