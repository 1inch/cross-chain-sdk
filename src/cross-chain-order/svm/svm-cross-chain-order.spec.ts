import {parseEther, parseUnits} from 'ethers'
import {SvmCrossChainOrder} from './svm-cross-chain-order'
import {NetworkEnum} from '../../chains'
import {TimeLocks} from '../../domains/time-locks'
import {HashLock} from '../../domains/hash-lock'
import {AuctionDetails} from '../../domains/auction-details'
import {EvmAddress, SolanaAddress} from '../../domains/addresses'

describe('SVMCrossChainOrder', () => {
    it('should correct calculate order hash', () => {
        const order = SvmCrossChainOrder.new(
            // 1 WETH [solana] -> 1000 USDC [ethereum]
            {
                maker: SolanaAddress.fromBigInt(1n),
                receiver: EvmAddress.fromBigInt(2n),
                srcToken: SolanaAddress.fromBigInt(3n),
                dstToken: EvmAddress.fromBigInt(4n), // address on dst chain
                srcAmount: parseEther('1'),
                minDstAmount: parseUnits('1000', 6)
            },
            {
                srcChainId: NetworkEnum.SOLANA,
                dstChainId: NetworkEnum.ETHEREUM,
                srcSafetyDeposit: 1000n,
                dstSafetyDeposit: 1000n,
                timeLocks: TimeLocks.fromDurations({
                    srcFinalityLock: 10n,
                    srcPrivateWithdrawal: 200n,
                    srcPublicWithdrawal: 100n,
                    srcPrivateCancellation: 100n,
                    dstFinalityLock: 10n,
                    dstPrivateWithdrawal: 100n,
                    dstPublicWithdrawal: 100n
                }),
                hashLock: HashLock.forSingleFill(
                    '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1'
                )
            },
            {
                auction: AuctionDetails.noAuction(120n, 1752739636n)
            },
            {
                allowMultipleFills: false,
                salt: 0x63030535n
            }
        )

        const hash = order.getOrderHashBuffer()

        expect(hash.toString('hex')).toEqual(
            'f59558e3ddb28b3aaa4953e12efc4c7b0d98c637942be3b5e8615998a5bf65c4'
        )
    })

    it('should correct toJSON/fromJSON', () => {
        const order = SvmCrossChainOrder.new(
            // 1 WETH [solana] -> 1000 USDC [ethereum]
            {
                maker: SolanaAddress.fromBigInt(1n),
                receiver: EvmAddress.fromBigInt(2n),
                srcToken: SolanaAddress.fromBigInt(3n),
                dstToken: EvmAddress.fromBigInt(4n), // address on dst chain
                srcAmount: parseEther('1'),
                minDstAmount: parseUnits('1000', 6)
            },
            {
                srcChainId: NetworkEnum.SOLANA,
                dstChainId: NetworkEnum.ETHEREUM,
                srcSafetyDeposit: 1000n,
                dstSafetyDeposit: 1000n,
                timeLocks: TimeLocks.fromDurations({
                    srcFinalityLock: 10n,
                    srcPrivateWithdrawal: 200n,
                    srcPublicWithdrawal: 100n,
                    srcPrivateCancellation: 100n,
                    dstFinalityLock: 10n,
                    dstPrivateWithdrawal: 100n,
                    dstPublicWithdrawal: 100n
                }),
                hashLock: HashLock.forSingleFill(
                    '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1'
                )
            },
            {
                auction: AuctionDetails.noAuction(120n, 1752739636n)
            },
            {
                allowMultipleFills: false,
                salt: 0x63030535n
            }
        )

        const json = order.toJSON()

        expect(json).toMatchSnapshot()
        expect(SvmCrossChainOrder.fromJSON(json)).toEqual(order)
    })
})
