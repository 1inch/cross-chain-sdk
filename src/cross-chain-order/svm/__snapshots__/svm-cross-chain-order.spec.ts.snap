// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SVMCrossChainOrder DstImmutables correct address type passed 1`] = `
{
  "amount": "1000000010",
  "hashlock": "0x11f8dd293e31c96da8ed3c186011fd0eb9036cce32f2e08c1c5fd5001b737907",
  "maker": "0x0000000000000000000000000000000000000002",
  "orderHash": "0xf59558e3ddb28b3aaa4953e12efc4c7b0d98c637942be3b5e8615998a5bf65c4",
  "safetyDeposit": "1000",
  "taker": "0x000000000000000000000000000000000000000a",
  "timelocks": "1318191364591968140515317923577708210251486390651022433320970",
  "token": "0x0000000000000000000000000000000000000004",
}
`;

exports[`SVMCrossChainOrder should correct toJSON/fromJSON 1`] = `
{
  "details": {
    "auction": {
      "duration": "120",
      "initialRateBump": 0,
      "points": [],
      "startTime": "1752739636",
    },
  },
  "escrowParams": {
    "dstChainId": 1,
    "dstSafetyDeposit": "1000",
    "hashLock": "0x11f8dd293e31c96da8ed3c186011fd0eb9036cce32f2e08c1c5fd5001b737907",
    "srcChainId": 501,
    "srcSafetyDeposit": "1000",
    "timeLocks": "1318191364591968140515317923577708210251486390651022433320970",
  },
  "extra": {
    "allowMultipleFills": false,
    "orderExpirationDelay": "12",
    "resolverCancellationConfig": {
      "cancellationAuctionDuration": 1,
      "maxCancellationPremium": "1",
    },
    "salt": "1661142325",
    "source": "sdk",
    "srcAssetIsNative": false,
  },
  "orderInfo": {
    "dstToken": "0x0000000000000000000000000000000000000004",
    "maker": "11111111111111111111111111111112",
    "minDstAmount": "1000000000",
    "receiver": "0x0000000000000000000000000000000000000002",
    "srcAmount": "1000000000000000000",
    "srcToken": "11111111111111111111111111111114",
  },
}
`;
